cpu/efm32/periph/i2c.c

```diff
@@ -54,28 +56,21 @@ static mutex_t i2c_lock[I2C_NUMOF];
  */
 static int _transfer(i2c_t dev, I2C_TransferSeq_TypeDef *transfer)
 {
-    bool busy = true;
-    uint32_t timeout = I2C_TIMEOUT;
+    uint32_t counter = EFM32_I2C_TIMEOUT;
 
-    /* start the i2c transaction */
-    i2c_progress[dev] = I2C_TransferInit(i2c_config[dev].dev, transfer);
+    /* start transfer (using interrupts) and wait for it to complete or
+       timeout */
+    NVIC_ClearPendingIRQ(i2c_config[dev].irq);
+    NVIC_EnableIRQ(i2c_config[dev].irq);
 
-    /* the transfer progresses via the interrupt handler */
-    while (busy) {
-        unsigned int cpsr = irq_disable();
+    i2c_progress[dev] = I2C_TransferInit(i2c_config[dev].dev, transfer);
 
-        if (i2c_progress[dev] == i2cTransferInProgress && timeout--) {
-            cortexm_sleep_until_event();
-        }
-        else {
-            busy = false;
-        }
+    while (i2c_progress[dev] == i2cTransferInProgress && counter--) {}
```

Comment:
```suggestion
    while (i2c_progress[dev] == i2cTransferInProgress && counter--) {
        thread_yield();
    }
```

Reply:
Tried that, but that may cause the thread to never continue.

Reply:
Or rather not - do we get an interrupt for each byte?
I guess `I2C_Transfer()` is an SDK function

Reply:
Yes. `I2C_Transfer` is an SDK function. An interrupt is raised for: `I2C_IEN_NACK | I2C_IEN_ACK | I2C_IEN_MSTOP | I2C_IEN_RXDATAV | I2C_IEN_ERRORS`.